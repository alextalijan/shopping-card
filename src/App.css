.header {
  background-color: #1e293b;
  color: white;
  height: 100px;
  display: flex;
  justify-content: end;
  align-items: center;
  gap: 2rem;
}

.nav-link {
  text-decoration: none;
  font-size: 2rem;
  transition: transform 0.3s;
}

.nav-link:hover {
  transform: scale(1.2);
}

.nav-link:visited {
  color: white;
}

.nav-link:last-child {
  margin-right: 2rem;
}
