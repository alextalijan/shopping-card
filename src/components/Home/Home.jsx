import { Link } from 'react-router-dom';
import styles from './Home.module.css';

function Home() {
  return (
    <>
      <h1 className={styles.h1}>Welcome to the All-Around Shop!</h1>
      <p className={styles['intro-text']}>
        Here you can buy anything from clothing to laptop bags.
      </p>
      <div className={styles['images-container']}>
        <img src="/public/shirt.png" alt="" />
        <img src="/public/laptop-bag.png" alt="" />
      </div>
      <p className={styles['shop-redirect-text']}>
        Click on the{' '}
        <Link className={styles['nav-link']} to="/shop" data-testid="shop-link">
          SHOP
        </Link>{' '}
        to browse our store.{' '}
      </p>
    </>
  );
}

export default Home;
