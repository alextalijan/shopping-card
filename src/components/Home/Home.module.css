.h1 {
  font-size: 4rem;
  text-align: center;
  margin: 4rem auto;
}

.intro-text {
  font-size: 2rem;
  text-align: center;
  opacity: 0.6;
  margin: 4rem auto;
}

.images-container {
  display: flex;
  justify-content: center;
  align-items: center;
  gap: 5rem;
}

.shop-redirect-text {
  font-size: 2rem;
  text-align: center;
  margin: 4rem auto;
}

.nav-link {
  text-decoration: none;
  font-size: 2rem;
  transition: transform 0.3s;
  display: inline-block;
  color: blueviolet;
}

.nav-link:hover {
  transform: scale(1.2);
}

.nav-link:visited {
  color: blueviolet;
}

@media (max-width: 1010px) {
  .h1 {
    font-size: 3.5rem;
    padding: 0 1rem;
  }

  .intro-text {
    font-size: 1.8rem;
    padding: 0 1rem;
  }

  .images-container {
    flex-direction: column;
    max-width: 100%;
  }

  .images-container > img {
    width: 300px;
  }

  .images-container :last-child {
    display: none;
  }
}
