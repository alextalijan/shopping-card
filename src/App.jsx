import { useEffect, useState } from 'react';
import './App.css';
import { Outlet, Link } from 'react-router-dom';

function App() {
  const [cart, setCart] = useState([]);
  const [shopItems, setShopItems] = useState([]);
  const [loadingShop, setLoadingShop] = useState(true);
  const [shopError, setShopError] = useState(null);

  function addToCart(id, title, imageSrc, price, amount) {
    if (amount === 0) return;

    const addedItem = {
      id,
      title,
      imageSrc,
      price,
      amount,
    };

    for (const item of cart) {
      if (addedItem.id === item.id) {
        const newCart = cart.map((item) => {
          if (item.id === addedItem.id) {
            return {
              id: addedItem.id,
              title: addedItem.title,
              price: addedItem.price,
              amount: item.amount + addedItem.amount,
            };
          }

          return item;
        });

        setCart(newCart);
        return;
      }
    }

    setCart([...cart, addedItem]);
  }

  function removeFromCart(id) {
    setCart(cart.filter((item) => item.id !== id));
  }

  useEffect(() => {
    fetch('https://fakestoreapi.com/products/')
      .then((response) => {
        if (!response.ok) {
          throw new Error(
            'Failed to load items from the shop. Please try again by refreshing the page.',
          );
        }

        return response.json();
      })
      .then((data) => {
        const items = [];

        for (let i = 0; i < 10; i += 1) {
          items.push({
            id: data[i].id,
            title: data[i].title,
            price: data[i].price,
            imageSrc: data[i].image,
          });
        }

        setShopItems(items);
      })
      .catch((error) => {
        setShopError(error.message);
      })
      .finally(() => {
        setLoadingShop(false);
      });
  }, []);

  return (
    <>
      <header className="header">
        <Link className="nav-link" to="/">
          Home
        </Link>
        <Link className="nav-link" to="/shop">
          Shop
        </Link>
        <Link className="nav-link" to="/cart">
          Cart <span>({cart.length})</span>
        </Link>
      </header>
      <main>
        <Outlet
          context={{
            shopItems,
            loadingShop,
            shopError,
            addToCart,
            cart,
            setCart,
            removeFromCart,
          }}
        />
      </main>
    </>
  );
}

export default App;
