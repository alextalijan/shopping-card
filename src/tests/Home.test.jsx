import { it, expect } from 'vitest';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { screen, render } from '@testing-library/react';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import Home from '../components/Home/Home';
import App from '../App';

it('renders', () => {
  const { container } = render(
    <MemoryRouter initialEntries={['/']}>
      <Routes>
        <Route path="/" element={<App />}>
          <Route index element={<Home />} /> {/* Home page */}
        </Route>
      </Routes>
    </MemoryRouter>,
  );
  expect(container).toMatchSnapshot();
});

it('switches to shop page after clicking on the link', async () => {
  const user = userEvent.setup();
  render(
    <MemoryRouter initialEntries={['/']}>
      <Routes>
        <Route path="/" element={<App />}>
          <Route index element={<Home />} /> {/* Home page */}
          <Route path="shop" element={<h1>Shop</h1>} /> {/* Shop page */}
        </Route>
      </Routes>
    </MemoryRouter>,
  );
  await user.click(screen.getByTestId('shop-link'));
  expect(screen.getByRole('heading', { name: /shop/i })).toBeInTheDocument();
});
