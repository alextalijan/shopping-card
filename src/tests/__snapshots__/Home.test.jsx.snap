// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders 1`] = `
<div>
  <header
    class="header"
  >
    <a
      class="nav-link"
      data-discover="true"
      href="/"
    >
      Home
    </a>
    <a
      class="nav-link"
      data-discover="true"
      href="/shop"
    >
      Shop
    </a>
    <a
      class="nav-link"
      data-discover="true"
      href="/cart"
    >
      Cart 
      <span>
        (
        0
        )
      </span>
    </a>
  </header>
  <main>
    <h1
      class="_h1_d6eec2"
    >
      Welcome to the All-Around Shop!
    </h1>
    <p
      class="_intro-text_d6eec2"
    >
      Here you can buy anything from clothing to laptop bags.
    </p>
    <div
      class="_images-container_d6eec2"
    >
      <img
        alt=""
        src="/public/shirt.png"
      />
      <img
        alt=""
        src="/public/laptop-bag.png"
      />
    </div>
    <p
      class="_shop-redirect-text_d6eec2"
    >
      Click on the
       
      <a
        class="_nav-link_d6eec2"
        data-discover="true"
        data-testid="shop-link"
        href="/shop"
      >
        SHOP
      </a>
       
      to browse our store.
       
    </p>
  </main>
</div>
`;
